### ISSUE TITLE
Track whether a field has been dirty since last initialize even if it currently isn't

### ISSUE THREAD

**Issue by connorwyatt** [redux-form/redux-form#3719]
### Are you submitting a **bug report** or a **feature request**?
Potentially both.

### What is the current behavior?
Currently, if I change the value of an unticked checkbox (input type is irrelevant, checkbox just used for simplicity) to true the pristine and dirty values get set to `true` and `false` (respectively). Then when I uncheck the checkbox again, the pristine and dirty values get set to `false` and `true` (respectively).

### What is the expected behavior?
I believe the standard definition of dirty is that the field value has changed, not that it is not equal to its original value, so therefore in this case the value of dirty should not change back to `false` as the value has been changed, it just happens to be what it originally was.

### Sandbox Link
https://codesandbox.io/s/5z9m63j21k

Note that changing the checkbox back to false makes the form pristine again.

### What's your environment?
Redux Form 7.1.2

### Other information
I'm aware that you may disagree with what I consider to be the standard definition of dirty, and am not just asking for it to be changed to another definition as that would cause a breaking change anyway. So at least for the short term it might make sense to just extend the API with a new property (e.g. `hasBeenChanged`) and its inverse (e.g. `hasNotBeenChanged`) rather than rework how the value of dirty is calculated.

Furthermore, if you are in agreement that the API changes are valuable, I am happy to create a PR. Just let me know your preference for the property name(s).

**Comment by connorwyatt** [comment]
Hey, sorry to bump this, but has this been seen? As said previously I would be happy to create a PR for this, however I'm reluctant to do so unless this feature/bug fix would be welcomed.

**Comment by connorwyatt** [comment]
Is there any update on this? It is a feature that would improve our UX significantly as we can only show errors when the field has been edited. Would just like to hear some thoughts on this

**Comment by bbenezech** [comment]
I guess you should use `meta.touched` instead.
Dirty means the field's value is not pristine.

**Comment by connorwyatt** [comment]
Touched isn't the same thing, touched is set to true when the field is focused and then defocused, regardless of whether the value was changed.

Here's the scenario:

A user is about to type in a field that is blank by default. We want to show the validation when they have edited the field, so we started to use the dirty property. That works fine, until the user deletes the contents of the field. `"" === ""` so the dirty prop is now false, meaning that no validation message to show the field is required is shown.

I this case, if we used touched, the user could focus a field and then defocus it, and validation messages show, something we want to avoid.

The bottom line is, the feature we need (an it is a common feature in a lot of form libraries) is a standard dirty implementation which just tracks if a change has been made to the value since the form was initialised. As I said before I appreciate that modifying the behaviour of the dirty property would be a breaking change, so it could be called something else, but that standard dirty behaviour is what is lacking at the moment.

Happy to create a PR if the concept is agreed upon.

**Comment by connorwyatt** [comment]
A couple of examples:

- [From the Angular docs](https://angular.io/api/forms/AbstractControl)
- [From a Vue.js library](https://www.npmjs.com/package/vue-form) (example [here](https://jsfiddle.net/fergal_doyle/bqys2p5y/))

**Comment by elliotaplant** [comment]
@connorwyatt - I'm not sure if I understand what you are asking for. When you say, "show the validation," do you mean "show errors if the field is invalid"?

I also want to clarify when you want to show errors. If the user focuses on the field and changes the value to something other than "", you want validation to show. If the user focuses on the field and changes it to some text but then changes it back to `""`, you want validation to show. If the user touches the field _but does not change it_ then you do _not_ want validation to show.

I personally wouldn't normally expect that last case if I was filling out a form (ie, if I touched a required field and without editing it, I would expect an error to show), but you could make an HOC that wraps the redux-form Field component to manage the `hasBeenChanged` state:

```
class HasBeenChangedField {
  constructor(props) {
    super(props);
    this.state = {
      hasBeenChanged: false,
    };
    this.handleOnChange = this.handleOnChange.bind(this);
  }

  handleOnChange(changeEvent) {
    if (!this.state.hasBeenChanged) {
      this.setState({ hasBeenChanged: true });
    }
  }

  render() {
    return (
      <Field
        onChange={this.handleOnChange}
        hasBeenChanged={this.state.hasBeenChanged}
        hasNotBeenChanged={!this.state.hasBeenChanged}
        {...this.props}
      />);
  }
}
```

**Comment by connorwyatt** [comment]
@elliotaplant What I'm asking for is in the initial post, I'm not sure how else to explain it. At its most basic, I think that the `dirty` implementation is not the idiomatic implementation of `dirty`, see the original post for more details.

As for the HOC, I respect your opinion, but our desired UX requires what I mentioned, and you're losing all state when the component is unmounted, which is not always desired. It's fine if you're doing a single page form, but sometimes you want the form to live longer than its fields. To counteract this you would need to implement your own reducer for tracking the state, and hook into the events dispatched by the redux form library, not really ideal.

**Comment by elliotaplant** [comment]
Great point - I assumed you would only need to know the `hasBeenTouched` state within your component, so it wouldn't matter if the state was lost on dismount. It seems like redux-form's implementation of `dirty` is really more like `changed`, and to match other form libraries it would set make `dirty` more like `hasBeenChanged`. 

I sympathize with you, but I doubt that change will happen. I think your best bet is to make a PR to add the `hasBeenChanged` state. Good luck!

**Comment by connorwyatt** [comment]
@elliotaplant Exactly, that's the behaviour that we're after. I'm totally up for creating a PR for this but haven't received the green light from someone on the project to say that it would be merged in, so I'm reluctant to waste time if it's going to be ignored. So fingers crossed someone responds to let me know it wouldn't be wasted time 😄
